### PR TITLE
Topic attribute bound electrons

### PR DESCRIPTION
**Pull request by n01r**

### CODE REVIEW

**Comment by n01r** [comment]
Hey @ax3l and @psychocoderHPC again,
This PR is the continuation of my previous PR #6 and is based on that branch. 
It changes the attribute `chargeState` to `boundElectrons`.
Feel free to comment!

File: src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSI.hpp [hunk 1]
```diff
@@ -56,13 +58,16 @@ namespace ionization
...
+            float_X protonNumber = precisionCast<float_X>(protonNumber_tmp);
+            float_X chargeState = attribute::getChargeState(parentIon);
             
             /* ionization condition */
-            if (math::abs(eField)*UNIT_EFIELD >= SI::IONIZATION_EFIELD && parentIon[chargeState_] < protonNumber)
+            if (math::abs(eField)*UNIT_EFIELD >= SI::IONIZATION_EFIELD && chargeState < protonNumber)
             {
                 /* set new particle charge state */
-                parentIon[chargeState_] = 1 + parentIon[chargeState_];
+                parentIon[boundElectrons_] = parentIon[boundElectrons_] - 1;
```
Comment (ax3l) [inline src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSI.hpp]: 1 is `int`, use `float_X(1.0)`

File: src/picongpu/include/particles/traits/GetProtonNumber.hpp [hunk 2]
```diff
@@ -33,7 +33,7 @@ struct GetProtonNumber
 {
     typedef typename PMacc::traits::Resolve<
         typename GetFlagType<typename T_Species::FrameType,protonNumber<> >::type 
-        >::type type;
+    >::type type;
```
Comment (ax3l) [inline src/picongpu/include/particles/traits/GetProtonNumber.hpp]: unnecessary change, revert pls.

File: src/picongpu/include/simulation_defines/param/speciesAttributes.param [hunk 3]
```diff
@@ -55,13 +55,13 @@ value_identifier(float_X, weighting, 0.0);
...
+/** number of electrons bound to the atom / ion
  *
  * value type is float_X to avoid casts during the runtime
- *
+ * - float is also reasonable because effective charge numbers are possible
  * - only reasonable for ion species if ionization is enabled 
- * - \todo maybe calculate from new attribute boundElectrons */
-value_identifier(float_X,chargeState,float_X(1.0));
+ */
+value_identifier(float_X,boundElectrons,float_X(1.0));
```
Comment (ax3l) [inline src/picongpu/include/simulation_defines/param/speciesAttributes.param]: Probably `1.0` is not the best "general" default value.
Maybe "fully ionized" aka `float_X(0.0)` is a better choice for now (or later on and alternatively, the proton-number for "neutral").

File: src/picongpu/include/traits/attribute/GetChargeState.hpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ * This is the fallback implementation if no `boundElectrons` are available for a particle
+ */
+template<>
+struct LoadChargeState<false>
+{
+
+    template<typename T_Particle>
+    HDINLINE float_X operator()(const T_Particle& particle)
+    {
+        return 1.0;
```
Comment (ax3l) [inline src/picongpu/include/traits/attribute/GetChargeState.hpp]: ? the default should be zero, too.

Comment (ax3l) [inline src/picongpu/include/traits/attribute/GetChargeState.hpp]: (pls stop assuming "the world is only neutral hydrogen" while coding ;) )

Comment (n01r) [inline src/picongpu/include/traits/attribute/GetChargeState.hpp]: That is not what I'm doing. The thing is that as long as ions get cloned out of electrons, I cannot start a simulation with neutral particles. If that particle has `boundElectrons` then I can calculate a `chargeState`. If it doesn't have `boundElectrons` it is most likely an electron and thus singly charged. I could also put an error message in there. "It is no use to ask for a charge state of particles that have no bound electrons because then they are probably elementary particles and their charge does not change. But if you still want to: default is 1 because `charge = CHARGE_CONST * chargeState`." The whole thing is just too preliminary. 
What should happen if you try to call `getChargeState` on a particle without differing charge states?

Comment (ax3l) [inline src/picongpu/include/traits/attribute/GetChargeState.hpp]: I understand your confusion, but as explained via mail:

> The thing is that as long as ions get cloned out of electrons ...

that is due to historic reasons, feel free to change that to what I described you: initialize all ions first, then clone the electrons from them.

Comment (n01r) [inline src/picongpu/include/traits/attribute/GetChargeState.hpp]: As I understood it yesterday @psychocoderHPC already has an unproposed branch that allows exactly this. I don't want to meddle in it before he starts a PR.

**Comment by ax3l** [comment]
that's fine so simply wait for that :)

else, it's a trivial change in `src/picongpu/include/simulation_defines/param/species*.param`

File: src/picongpu/include/traits/attribute/GetChargeState.hpp (hunk 4 above)
Comment (ax3l) [inline src/picongpu/include/traits/attribute/GetChargeState.hpp]: that's fine so simply wait for that :)

else, it's a trivial change in `src/picongpu/include/simulation_defines/param/species*.param`

**Comment by psychocoderHPC** [comment]
A general question: 
I don't know if is plays any roll but is `freeElectrons` a more native attribute instead of `boundElectrons`?
The advantage of `freeElectrons` is that for all neutral materials the default is zero.

**Comment by n01r** [comment]
@psychocoderHPC 
If you would rename it like that it is the same as `chargeState` just with a different name. That would be zero, too for all neutral materials. It is a bit more native than `chargeState` but as we agreed on before, `boundElectrons` can be extended easier to coding what energy states the bound electrons are actually in.

**Comment by ax3l** [comment]
FreeElectrons does not make sence since these e- are not associated with the ion any more. So no :)

**Comment by psychocoderHPC** [comment]
yes this make sense
